File: checks/python/src/section508/checks/status_messages.py
import re

from section508.dom import attr, select

MESSAGE = re.compile(r"saved|error|items in cart|loading|success|thank you|updated|complete", re.I)


def check(root):
    if not root:
        return []

    messages = [element for element in select(root, "div, p, output, span") if MESSAGE.search(element.get_text() or "")]
    if not messages:
        return ["This section has no status message."]

    for element in messages:
        role = (attr(element, "role") or "").lower()
        live = (attr(element, "aria-live") or "").lower()
        if role in {"status", "alert"} or live in {"polite", "assertive"}:
            continue
        return ["The status message is not a live region."]
    return []

File: checks/ruby/lib/section508/checks/status_messages.rb
# frozen_string_literal: true

module Section508
  module Checks
    # A confirmation or progress line needs a live region.
    class StatusMessages
      MESSAGE = /saved|error|items in cart|loading|success|thank you|updated|complete/i

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        messages = Dom.elements(root, "div, p, output, span").select do |element|
          MESSAGE.match?(element.text.to_s)
        end
        return ["This section has no status message."] if messages.empty?

        messages.each do |element|
          role = element["role"].to_s.downcase
          live = element["aria-live"].to_s.downcase
          next if %w[status alert].include?(role) || %w[polite assertive].include?(live)

          return ["The status message is not a live region."]
        end
        []
      end
    end
  end
end
